posts/zzz_DO_NOT_EDIT_the__tensio.../the__tension_of__high-_performance__computing:__reproducibility_vs.__parallelization.qmd

```diff
@@ -0,0 +1,259 @@
+across multiple workers.
+
+```{r, message=FALSE, eval=FALSE}
+library("mirai")
+library("dplyr", warn.conflicts = FALSE)
+
+# start parallel workers
+daemons(4)
+
+# load libraries on each worker and set up logging to a file
+everywhere(
+  {
+    library("dplyr")
+    library("tidyr")
+    library("mirai")
+    library("tidylog", warn.conflicts = FALSE)
+
+    # Define function to log messages to the log file
+    log_to_file <- \(txt) cat(txt, file = log_file, sep = "\n", append = TRUE)
+    options("tidylog.display" = list(message, log_to_file))
+  },
+  log_file = "log.txt"
+)
+
+# perform computations in parallel
+m <- mirai_map(letters[1:3], \(x) {
+  mutate(tibble(.rows = 1), "{x}" := "foo")
+})
+
+# collect results
+result <- m[] |> dplyr::bind_cols()
+
+daemons(0)
+
+return(result)
+```
+
+In the above code chunk, we set up a parallel processing environment using the
+`{mirai}` package. The function `mirai_map()` is used to apply a mutating
+function in parallel to a tibble for each element of `letters[1:3]`, logging the
+operations to a file using the `{tidylog}` package. However, while we can log
+each operation as it happens, due to the parallel nature of `{mirai}`, the 
+logging does not occur in a controlled or sequential order. *Each daemon 
+executes its task independently, and the order of logging in the file will
+depend on the completion times of these parallel processes rather than the
+intended flow of operations.*
+
+> Parallel computations can obscure the traceability of operations
+
+This lack of control can lead to a situation where the log entries do not
+reflect the actual sequence in which the `{dplyr}` commands were expected to be
+processed. Although the operations themselves are carried out correctly, the
+asynchronous logging may create challenges in *tracing* and *debugging* the
+process, as entries in the log file could appear out of order, giving an
+incomplete or misleading representation of the task flow.
+
+```{r}
+readLines("log.txt")
+```
+
+In the above code, when we read the contents of the log file, you will notice
+that the logs are not in the same order as the commands were dispatched. This
+demonstrates the inherent difficulty in managing the order of logging in
+parallel tasks, especially when there is no guarantee on how quickly each
+process will complete and record its operations. 
+
+### Task Dispatching and RNG Management
+
+By default, `{mirai}` uses an advanced dispatcher to manage task distribution
+efficiently, scheduling tasks in a First-In-First-Out manner and leveraging
+[`{nanonext}`](https://shikokuchuo.net/nanonext/) primitives for zero-latency,
+resource-free task management. However, its asynchronous execution can hinder
+reproducibility, especially with random number generation (RNG) or tasks needing
+strict order.
+
+To enhance reproducibility, `{mirai}` allows disabling the dispatcher, directly
+connecting the host to daemons in a round-robin fashion. While less efficient, 
```

Comment:
what is: round-robin?

Reply:
Good catch, I rewrote this simpler - and added a link to the underlying algorithm.